Use mavenLocal() *first* for repositories in Gradle integration tests

yrodiere opened the pull request:
This solves random failures we've been seeing when trying to fetch
snapshots of Quarkus artifacts from the Gradle plugin portal (see #26126, #26128). When executing the builds locally with --debug to reproduce the errors, I spotted this:

```
    Caused by: org.gradle.internal.resource.transport.http.HttpErrorStatusCodeException: Could not HEAD 'https://repo.gradle.org/artifactory/jcenter/io/quarkus/quarkus-build-parent/999-SNAPSHOT/quarkus-build-parent-999-SNAPSHOT.pom'. Received status code 409 from server: Conflict
```

I don't know why this repository would return HTTP 409 conflict, but the
more important problem is that we shouldn't be looking for this artifact
in this repository in the first place.

If we use mavenLocal(), we might as well look into the maven local
repository *first*.

Indeed, in these integration tests, a Gradle plugin or a dependency
is always either:

- a snapshot dependency, in which case it's part of the Quarkus build
  and we just built it, so we really want to fetch it from the Maven
  local repository.
- or a non-snapshot dependency, in which case the artifact we find in the
  Maven local repository (if any) will always be exactly the one we're
  looking for; and if it's not there we'll fall back to Maven Central
  or the Gradle plugin Portal anyway.

yrodiere committed:
Use mavenLocal() *first* for repositories in Gradle integration tests

This solves random failures we've been seeing when trying to fetch
snapshots of Quarkus artifacts from the Gradle plugin portal:

```
    Caused by: org.gradle.internal.resource.transport.http.HttpErrorStatusCodeException: Could not HEAD 'https://repo.gradle.org/artifactory/jcenter/io/quarkus/quarkus-build-parent/999-SNAPSHOT/quarkus-build-parent-999-SNAPSHOT.pom'. Received status code 409 from server: Conflict
```

I don't know why this repository would return HTTP 409 conflict, but the
more important problem is that we shouldn't be looking for this artifact
in this repository in the first place.

If we use mavenLocal(), we might as well look into the maven local
repository *first*.

Indeed, in these integration tests, a Gradle plugin or a dependency
is always either:

- a snapshot dependency, in which case it's part of the Quarkus build
  and we just built it, so we really want to fetch it from the Maven
  local repository.
- or a non-snapshot dependency, in which case the artifact we find in the
  Maven local repository (if any) will always be exactly the one we're
  looking for; and if it's not there we'll fall back to Maven Central
  or the Gradle plugin Portal anyway.

yrodiere commented:
Note this did solve the problem locally, but I'm no Gradle expert and I don't know what the good practices are regarding `mavenLocal()`. Apparently some people say the good practice is not to use it at all, but that's not helpful because in our case (mix of Maven and Gradle in the same build) we *have* to use it.

gsmet commented:
It was changed recently the other way so I don't know but given we start seeing random Gradle failures, we probably need a fix of some sort.

/cc @glefloch @sebersole

yrodiere commented:
> It was changed recently

Ah, right, it was changed the other way around... but I see the order was updated as far back as January/May 2021. Were there more recent changes?

Definitely worth some input from @glefloch.

yrodiere commented:
> Are you talking specifically about resolving plugins via `mavenLocal` during testing or resolving artifacts from `mavenLocal` while testing the plugin?

Mostly when resolving plugins, as I encountered the problem while Gradle was resolving plugins (and their dependencies, some of which happen to be simply Quarkus artifacts). But I imagine we could have the same problem when resolving artifacts, since test projects usually depend on Quarkus extensions.

>  It would be much better to do that through Gradle and use its "TestKit" approach.

I'll trust you on that; no idea how that works. If this PR works, that could be enough for now, though? I just want to get the build unstuck so I can have a passing CI build for my other PRs :)

yrodiere commented:
Ok, now it's failing for different tests, and in a different way:

```
2022-06-15T09:51:42.1219095Z Execution failed for task ':ext-a:runtime:extensionDescriptor'.
2022-06-15T09:51:42.1219732Z > Could not find jakarta.xml.bind-api-2.3.3.jar (jakarta.xml.bind:jakarta.xml.bind-api:2.3.3).
2022-06-15T09:51:42.1220132Z   Searched in the following locations:
2022-06-15T09:51:42.1220823Z       file:/home/runner/.m2/repository/jakarta/xml/bind/jakarta.xml.bind-api/2.3.3/jakarta.xml.bind-api-2.3.3.jar
```

What's weird is that the dependency was only resolved from the first repository (in this case `mavenLocal()`). I suspect there's something in the gradle extension that resolves dependencies (maybe for deployment artifacts) but does not correctly take into account the repository config?

I'm still digging...

yrodiere committed:
Simplify uses of mavenLocal() in Gradle integration tests

According to the Gradle documentation, mavenLocal() will take
maven.repo.local into account automatically:

>  The location for the repository is determined as follows (in order of precedence):
>
>    The value of system property 'maven.repo.local' if set;
>    The value of element <localRepository> of ~/.m2/settings.xml if this file exists and element is set;
>    The value of element <localRepository> of $M2_HOME/conf/settings.xml (where $M2_HOME is the value of the environment variable with that name) if this file exists and element is set;
>    The path ~/.m2/repository.

yrodiere committed:
Use mavenLocal() only for Quarkus artifacts in Gradle integration tests

yrodiere commented:
I pushed a few more commits to make sure mavenLocal() is only used to resolve Quarkus artifacts; let's see how it goes...

glefloch commented:
Sorry for the late answer, we shouldn't set `mavenLocal()` as the first repository because the local repository can be incomplete. 
The failure on missing jar for `jakarta.xml.bind:jakarta.xml.bind-api:2.3.3` exposes this issue. 

When resolving dependencies, Gradle will look into repository sequentially. As soon as it found the expected artifact coordinates (group:artifact:version), it expects all artifact types to be there however, in some case, the jar can be missing (for example when maven only resolves the pom). In that case instead of going to the next repository it throws an exception;
By default using `mavenLocal()` should be avoided or limited to sharing artifacts between a local maven project and a gradle project. Gradle handle cache in a specific `.gradle/caches` directory. 

I'm looking into your error which looks related to a really slow resolution time, and then lead to the timeout. I restarted the pipeline to check if the error is still here. I think we should leverage the github cache API, that could help with dependency resolution.

yrodiere commented:
> Sorry for the late answer, we shouldn't set `mavenLocal()` as the first repository because the local repository can be incomplete.

By "incomplete" I suppose you mean for example Maven downloaded the pom, but not the JAR. That's true.

But:

1. You can configure `mavenLocal()` to only be used to look up Quarkus artifacts / snapshots.
2. Putting `mavenLocal()` last *also* has downsides, which we're seeing right now with CI builds failing all over: Gradle will look up Quarkus artifacts in remote repositories first, which may lead to irrelevant errors like the 409 Conflict we're seing. In short we're relying even more on external infra, and that external infra may fail.

I just tried configuring `mavenLocal()` correctly, but unfortunately I'm still having (different) problems... it would seem that dependency resolution in the `extensionDescriptor` task does not follow exactly the same rules; in particular I'm getting that kind of error:

```
> Could not resolve all dependencies for configuration ':ext-a:runtime:runtimeClasspath'.
   > Could not find org.acme:simple-dependency:1.0-SNAPSHOT.
     Searched in the following locations:
       - https://repo.maven.apache.org/maven2/org/acme/simple-dependency/1.0-SNAPSHOT/maven-metadata.xml
       - https://repo.maven.apache.org/maven2/org/acme/simple-dependency/1.0-SNAPSHOT/simple-dependency-1.0-SNAPSHOT.pom
     Required by:
         project :ext-a:runtime
```

Which means it's ignoring `mavenLocal()` (it's not listed in the "following locations"), even though I have this configuration:

```
repositories {
    mavenLocal {
        content {
            includeGroup 'org.acme' // for dependencies built in this test
            includeGroupByRegex 'io.quarkus.*'
        }
    }
    mavenCentral()
}
```

I'm still trying to find out what's going on...

yrodiere commented:
Ok, I found the culprit: repositories were defined both in the root build.gradle and in the build.gradle of subprojects, and I had only updated the root. Now that I updated everything, the build passes.

@glefloch Could you please have a look at the changes? They do solve the build failures, they include a few cleanups, and with my last commit in particular, I think they comply with your requirement of "By default using `mavenLocal()` should be avoided **or limited to sharing artifacts between a local maven project and a gradle project**".

We can of course improve on this, in particular to use the gradle cache for internal dependencies between Gradle projects in `ConditionalDependenciesTest`, but I think this can be a follow-up PR?

glefloch commented:
Yes sure limiting `mavenLocal()` usage to quarkus artifact is really nice.

glefloch commented on integration-tests/gradle/src/main/resources/add-extension-multi-module-kts/build.gradle.kts:
```diff
@@ -1,10 +1,10 @@
 buildscript {
 
     repositories {
-        if (System.getProperties().containsKey("maven.repo.local")) {
```
I think we still need this condition, when building a release, a new local repository is used. And we used to face error in that case.

yrodiere replied:
That's strange. The gradle docs states that `maven.repo.local` is taken into account automatically in `mavenLocal()` ([source](https://docs.gradle.org/current/dsl/org.gradle.api.artifacts.dsl.RepositoryHandler.html#org.gradle.api.artifacts.dsl.RepositoryHandler:mavenLocal()))

>  The location for the repository is determined as follows (in order of precedence):
>
> *    **The value of system property 'maven.repo.local' if set;**
> *    The value of element <localRepository> of ~/.m2/settings.xml if this file exists and element is set;
> *    The value of element <localRepository> of $M2_HOME/conf/settings.xml (where $M2_HOME is the value of the environment variable with that name) if this file exists and element is set;
> *    The path ~/.m2/repository.

I just checked older docs and it's been that way forever, since at least 4.0.

Are you sure there wasn't another problem that was since solved? E.g. I see you had to add code in multiple places to propagate the system property from one build to another (from Maven to Gradle, but also sometimes from one Gradle build to another).

I just tested locally with the following command executed from `integration-tests/gradle/target/classes/conditional-dependencies`:

```
~/workspaces/main/quarkus/integration-tests/gradle/gradlew -Dmaven.repo.local=$HOME/.m2/repository --refresh-dependencies --no-build-cache -Dorg.gradle.console=plain --stacktrace --info :ext-a:runtime:publishToMavenLocal :ext-a:deployment:publishToMavenLocal :ext-b:runtime:publishToMavenLocal :ext-b:deployment:publishToMavenLocal :ext-c:runtime:publishToMavenLocal :ext-c:deployment:publishToMavenLocal :ext-d:runtime:publishToMavenLocal :ext-d:deployment:publishToMavenLocal :ext-e:runtime:publishToMavenLocal :ext-e:deployment:publishToMavenLocal :ext-f:runtime:publishToMavenLocal :ext-f:deployment:publishToMavenLocal :ext-g:runtime:publishToMavenLocal :ext-g:deployment:publishToMavenLocal :ext-h:runtime:publishToMavenLocal :ext-h:deployment:publishToMavenLocal :ext-i:runtime:publishToMavenLocal :ext-i:deployment:publishToMavenLocal :ext-j:runtime:publishToMavenLocal :ext-j:deployment:publishToMavenLocal :ext-k:runtime:publishToMavenLocal :ext-k:deployment:publishToMavenLocal :ext-l:runtime:publishToMavenLocal :ext-l:deployment:publishToMavenLocal :ext-m:runtime:publishToMavenLocal :ext-m:deployment:publishToMavenLocal :ext-n:runtime:publishToMavenLocal :ext-n:deployment:publishToMavenLocal :ext-o:runtime:publishToMavenLocal :ext-o:deployment:publishToMavenLocal :ext-p:runtime:publishToMavenLocal :ext-p:deployment:publishToMavenLocal :ext-r:runtime:publishToMavenLocal :ext-r:deployment:publishToMavenLocal :ext-s:runtime:publishToMavenLocal :ext-s:deployment:publishToMavenLocal :ext-t:runtime:publishToMavenLocal :ext-t:deployment:publishToMavenLocal :ext-u:runtime:publishToMavenLocal :ext-u:deployment:publishToMavenLocal
```

As is, it works, but if I change `maven.repo.local` to some non-existing repository, e.g. `$HOME/.m2/repository3`, it fails because it doesn't find Quarkus SNAPSHOT artifacts. So, evidently, the system property `maven.repo.local` **is** taken into account...

Maybe let's try? @gsmet can reach out to me very quickly if it causes a problem during a release, and he won't hesitate to put me to work :)

gsmet replied:
I think the CI release job we have set up should catch this (probably).

Also, we aren't running the tests anymore during the release process.

glefloch replied:
ok thanks

glefloch reviewed (changes requested):
This is a really nice work! I just think we still need the condition for custom maven repository used for building release (e.g @gsmet)

yrodiere commented:
Glad to see this merged :) Thanks @glefloch @gsmet @sebersole!
